huffman/test/decode_test.cpp

```diff
@@ -2,70 +2,277 @@
 
 #include <boost/ut.hpp>
 
+#include <algorithm>
 #include <array>
+#include <climits>
 #include <cstddef>
 #include <stdexcept>
 #include <utility>
 
-constexpr auto reverse_bits(int b) -> std::byte
+namespace {
+
+static constexpr auto eot = '\4';
+
+static constexpr auto code_table = [] {
+  using namespace ::starflate::huffman::literals;
+
+  // clang-format off
+  return ::starflate::huffman::table{
+      ::starflate::huffman::table_contents,
+      {
+          std::pair{0_c, 'e'},
+                   {10_c, 'i'},
+                   {110_c, 'n'},
+                   {1110_c, 'q'},
+                   {11110_c, eot},
+                   {11111_c, 'x'},
+      }};
+  // clang-format on
+}();
+
+}
+
+namespace test {
+
+constexpr auto verify(bool cond)
 {
-  std::byte result{};
-  for (auto i = 0; i < CHAR_BIT; ++i) {
-    result <<= 1;
-    result |= std::byte{(b & 1) == 1};
-    b >>= 1;
+  if (not cond) {
+    throw std::runtime_error{""};
   }
-  return result;
 }
 
+}  // namespace test
+
 auto main() -> int
 {
+  using ::boost::ut::eq;
   using ::boost::ut::expect;
   using ::boost::ut::test;
 
   namespace huffman = ::starflate::huffman;
-  using namespace huffman::literals;
-
-  // FIXME table contents are not in canonical form
-  test("basic") = [] {
-    // encoded data from soxofaan/dahuffman readme.rst.
-    // We reverse the bits in each byte to match the encoding used in DEFLATE.
-    constexpr std::array encoded_bytes = {
-        reverse_bits(134),
-        reverse_bits(124),
-        reverse_bits(37),
-        reverse_bits(19),
-        reverse_bits(105),
-        reverse_bits(64)};
-
-    constexpr char eot = {'\4'};
-    static constexpr auto code_table =  // clang-format off
-      huffman::table{
-        huffman::table_contents,
-        {std::pair{1_c,     'e'},
-                  {01_c,    'i'},
-                  {001_c,   'n'},
-                  {0001_c,  'q'},
-                  {00000_c, eot},
-                  {00001_c, 'x'}
-        }
-      };  // clang-format on
-
-    constexpr std::array expected = {
+
+  test("empty") = [] {
+    constexpr auto result = [] {
+      auto buf = std::array<char, 1>{};
+
+      auto it = huffman::decode(code_table, huffman::bit_span{}, buf.begin());
+
+      return it == buf.begin();
+    }();
+
+    static_assert(result);
+  };
+
+  test("just `e` - beginning not byte aligned") = [] {
+    static constexpr auto expected = std::array{'e'};
+
+    constexpr auto decoded = [] {
+      constexpr auto encoded = std::array<std::byte, 1>{};
```

Comment:
Zero initialized I assume?
Might be clearer to write the zero explicitly.

Reply:
It default constructs the array with default constructed `T`s.

Sure I can do that.